### PR TITLE
Fix TaskHost crash when task returns string[] with null elements

### LINKED ISSUE #13174: Taskhost should correctly serialize `string[]` containing null elements
## Summary

MSBuild's `-mt` mode crashes when a task returns a `String[]` containing null elements. The `BinaryTranslator` serialization code doesn't handle null array elements.

## Stack Trace

```
System.ArgumentNullException: Value cannot be null. (Parameter 'value')
   at System.IO.BinaryWriter.Write(String value)
   at Microsoft.Build.BackEnd.BinaryTranslator.BinaryWriteTranslator.Translate(String[]& array)
   at Microsoft.Build.BackEnd.TaskParameter.TranslatePrimitiveTypeArray(ITranslator translator)
   at Microsoft.Build.BackEnd.TaskHostTaskComplete.Translate(ITranslator translator)
   at Microsoft.Build.BackEnd.NodeEndpointOutOfProcBase.RunReadLoop(...)
```

## Reproduction

Any task returning `String[]` with null elements crashes in `-mt` mode:

```csharp
public class MyTask : Task
{
    [Output]
    public string[] Results { get; set; }

    public override bool Execute()
    {
        // Pre-allocated array not fully filled
        Results = new string[10];
        Results[0] = "first";
        // Results[1..9] are null - CRASH in -mt mode
        return true;
    }
}
```

## Risk Assessment

**High risk for custom tasks** - Common patterns trigger this:
- Pre-allocated arrays not fully filled
- LINQ projections: `items.Select(x => x?.Value).ToArray()`
- Sparse arrays for optional results

A `String[]` with null elements is valid .NET; MSBuild should serialize it correctly.

# Fix plan proposal: TaskHost should correctly serialize `string[]` containing null elements

## Problem Statement

MSBuild's `-mt` mode TaskHost crashes when a task returns a `String[]` containing null elements. The `BinaryTranslator` serialization code calls `BinaryWriter.Write(null)` which throws `ArgumentNullException`.

**Stack Trace:**
```
System.ArgumentNullException: Value cannot be null. (Parameter 'value')
   at System.IO.BinaryWriter.Write(String value)
   at Microsoft.Build.BackEnd.BinaryTranslator.BinaryWriteTranslator.Translate(String[]& array)
   at Microsoft.Build.BackEnd.TaskParameter.TranslatePrimitiveTypeArray(ITranslator translator)
   at Microsoft.Build.BackEnd.TaskHostTaskComplete.Translate(ITranslator translator)
```

**Real-World Impact:** The SDK's `GenerateGlobalUsings` task and similar tasks can trigger this bug. When building aspnetcore with `-mt` mode, TaskHost worker processes crash.

### Why TaskHost is affected but normal worker nodes are not

1. **TaskHost path (VULNERABLE):** Task outputs are serialized via `TaskHostTaskComplete` → `TaskParameter` → `BinaryTranslator.Translate(ref string[] array)`. This directly calls `BinaryWriter.Write(array[i])` without null checking.

2. **Normal worker path (SAFE):** Task outputs flow through in-process `IBuildEngine` callbacks that filter/convert values before any binary serialization occurs. The raw task output objects with null array elements never reach the serialization layer.

### Root Cause

In `BinaryTranslator.cs`, the `BinaryWriteTranslator.Translate(ref string[] array)` method (lines 1159-1173) iterates over the array and calls `_writer.Write(array[i])` without checking for null elements:

```csharp
for (int i = 0; i < count; i++)
{
    _writer.Write(array[i]);  // BinaryWriter.Write(null) throws!
}
```

Same issue affects:
- `List<string>` (lines 1189-1192)
- `HashSet<string>` (lines 1206-1208)

## Proposed Solution

Add null element handling by writing a boolean marker before each string element:

**Write side:**
```csharp
for (int i = 0; i < count; i++)
{
    string element = array[i];
    bool hasValue = element != null;
    _writer.Write(hasValue);
    if (hasValue)
    {
        _writer.Write(element);
    }
}
```

**Read side:**
```csharp
for (int i = 0; i < count; i++)
{
    bool hasValue = _reader.ReadBoolean();
    if (hasValue)
    {
        array[i] = _reader.ReadString();
    }
    // else: array[i] remains null (default)
}
```

### Scope of Changes

**Analysis confirms:** Task outputs are limited to scalar types and arrays (`string[]`, value type arrays, `ITaskItem[]`) per `TaskParameterTypeVerifier`. 

The `List<string>` and `HashSet<string>` direct translation methods in `BinaryTranslator` are **not actually used anywhere** in the codebase. All `List<string>` usages go through the generic `Translate<T>(ref List<T>, ObjectTranslator<T>)` overload, which calls `Translate(ref string)` per element - and that method already handles null via `TranslateNullable()`.

**Therefore, we only need to fix `string[]`** - the only collection type affected by this bug.

Files to modify:
1. `src/Framework/BinaryTranslator.cs` - Fix serialization for:
   - `BinaryWriteTranslator.Translate(ref string[] array)`
   - `BinaryReadTranslator.Translate(ref string[] array)`

2. `src/Shared/UnitTests/TaskParameter_Tests.cs` - Add test for null elements in string arrays

3. `src/Build.UnitTests/BackEnd/TaskHostTaskComplete_Tests.cs` - Add end-to-end test for the TaskHost scenario

## Work Plan

- [ ] **1. Add unit test for string[] with null elements in TaskParameter_Tests.cs**
  - Test that `TaskParameter` correctly serializes and deserializes a `string[]` with null elements
  - Similar pattern to existing `PrimitiveArrayParameter` test

- [ ] **2. Add end-to-end test in TaskHostTaskComplete_Tests.cs**
  - Test that `TaskHostTaskComplete` with string array output containing nulls serializes correctly

- [ ] **3. Fix BinaryWriteTranslator.Translate(ref string[] array)**
  - Add null check before writing each element
  - Write boolean marker to indicate presence of value

- [ ] **4. Fix BinaryReadTranslator.Translate(ref string[] array)**
  - Read boolean marker before reading each element
  - Only read string if marker indicates non-null

- [ ] **5. Run unit tests to verify the fix**
  - Run TaskParameter_Tests
  - Run TaskHostTaskComplete_Tests

- [ ] **6. Run broader test suite to ensure no regressions**
  - Build and run tests for the affected projects

### Binary Format Compatibility

This change modifies the wire format of serialized string arrays. However:
- TaskHost communication is within a single MSBuild process family (parent + task host children)
- All processes are from the same MSBuild version
- No cross-version compatibility concerns for this specific serialization path

### Performance

- Adds 1 byte per string element (boolean marker)
- Minimal impact on serialization performance
- Hot path performance is maintained

### Why only string[] needs fixing

Task outputs are restricted to scalar types and arrays per `TaskParameterTypeVerifier`. The `List<string>` and `HashSet<string>` translation methods are only used for internal MSBuild serialization (e.g., `SdkResult`), which don't contain null elements. Only `string[]` task outputs can have nulls from user tasks.

### Alternative Approaches Considered

1. **Filter nulls before serialization:** Would change semantics; task might depend on null positions
2. **Use different encoding for null (e.g., empty string marker):** Ambiguous with actual empty strings
3. **Throw earlier with descriptive error:** Doesn't fix the underlying limitation

The boolean marker approach is the cleanest solution that preserves null element semantics.

### PR DESCRIPTION
**Pull request by JanProvaznik**
Fixes #13174

### Context

When a task running in TaskHost (out-of-process task execution) returns a \string[]\ containing null elements, the build crashes with \ArgumentNullException\ in \BinaryTranslator.Translate(ref string[])\ because \BinaryWriter.Write(null)\ throws.
 
affects TaskHost scenarios. In-process task execution already filters null elements when converting to items in `TaskExecutionHost.GatherArrayStringAndValueOutputs()`.

### Changes Made

- Added \FilterNullsFromStringArray()\ helper in \OutOfProcTaskAppDomainWrapperBase.cs\ that filters null elements from string[] task outputs before serialization
- Logs a \MessageImportance.Normal\ message when nulls are filtered (for diagnostic purposes)
- Performance-efficient: single pass to count nulls, only allocates if nulls exist
- Added resource string \TaskHostAcquired_NullsFiltered\ for the log message

### Testing

end to end test with a task running in taskhost and producing null containing string array

### Notes

The fix produces identical results to in-process execution since MSBuild already discards null elements when converting task output arrays to items.

### CODE REVIEW

File: src/MSBuild/OutOfProcTaskAppDomainWrapperBase.cs [hunk 1]
```diff
@@ -419,7 +420,15 @@ private OutOfProcTaskHostTaskResult InstantiateAndExecuteTask(
                 {
                     try
                     {
-                        finalParameterValues[value.Name] = value.GetValue(wrappedTask, null);
+                        object outputValue = value.GetValue(wrappedTask, null);
+
+                        // Filter null elements from string[] outputs. They would be filtered anyway
```
Comment (JanProvaznik) [inline src/MSBuild/OutOfProcTaskAppDomainWrapperBase.cs]: improve the comment by linking to the issue, and saying less implementation details

File: src/MSBuild/OutOfProcTaskAppDomainWrapperBase.cs [hunk 2]
```diff
@@ -419,7 +420,15 @@ private OutOfProcTaskHostTaskResult InstantiateAndExecuteTask(
...
+                        // when converting to items (see TaskExecutionHost.GatherArrayStringAndValueOutputs),
+                        // but BinaryTranslator.Translate(ref string[]) crashes on null elements because
+                        // BinaryWriter.Write(null) throws ArgumentNullException.
+                        outputValue = FilterNullsFromStringArray(outputValue, value.Name);
```
Comment (JanProvaznik) [inline src/MSBuild/OutOfProcTaskAppDomainWrapperBase.cs]: wrap this with the string array check instead of doing it in the method

File: src/MSBuild/OutOfProcTaskAppDomainWrapperBase.cs [hunk 3]
```diff
@@ -451,5 +460,50 @@ private void LogErrorDelegate(string taskLocation, int taskLine, int taskColumn,
...
+                    break;
+                }
+            }
+
+            if (!hasNulls)
+            {
+                return outputValue;
+            }
+
+            // Filter nulls using LINQ and log
```
Comment (JanProvaznik) [inline src/MSBuild/OutOfProcTaskAppDomainWrapperBase.cs]:
```suggestion
            // Filter nulls and log
```

File: src/MSBuild/OutOfProcTaskAppDomainWrapperBase.cs [hunk 4]
```diff
@@ -451,5 +460,50 @@ private void LogErrorDelegate(string taskLocation, int taskLine, int taskColumn,
...
+            string[] filtered = stringArray.Where(s => s != null).ToArray();
+            int nullCount = stringArray.Length - filtered.Length;
+
+            buildEngine.LogMessageEvent(new BuildMessageEventArgs(
+                message: ResourceUtilities.FormatResourceStringIgnoreCodeAndKeyword("TaskHostAcquired_NullsFiltered", parameterName, nullCount),
+                helpKeyword: null,
+                senderName: taskName,
+                importance: MessageImportance.Low));
```
Comment (JanProvaznik) [inline src/MSBuild/OutOfProcTaskAppDomainWrapperBase.cs]: medium

File: src/MSBuild/OutOfProcTaskAppDomainWrapperBase.cs [hunk 5]
```diff
@@ -451,5 +460,50 @@ private void LogErrorDelegate(string taskLocation, int taskLine, int taskColumn,
...
+        private object FilterNullsFromStringArray(object outputValue, string parameterName)
+        {
+            if (outputValue is not string[] stringArray)
+            {
+                return outputValue;
+            }
+
+            // Quick check: does array contain any nulls?
+            bool hasNulls = false;
+            foreach (string s in stringArray)
```
Comment (JanProvaznik) [inline src/MSBuild/OutOfProcTaskAppDomainWrapperBase.cs:472]: just use .Contains()

**Comment by JanProvaznik** [comment]
we generally avoid linq
~~- I decided to use it because this is not a hot path (and the logging overhead dominates anyway)
- I realized that Contains is also linq thing and I should use indexof instead
~~ I refactored
filtering under the try: seems most logical to me to do it between the getter and setter, it would be confusing to do it after with managing the local variable scopes differently

File: src/MSBuild/OutOfProcTaskAppDomainWrapperBase.cs [hunk 6]
```diff
@@ -451,5 +460,46 @@ private void LogErrorDelegate(string taskLocation, int taskLine, int taskColumn,
...
+                {
+                    filtered[j++] = s;
+                }
+            }
+
+            buildEngine.LogMessageEvent(new BuildMessageEventArgs(
+                message: ResourceUtilities.FormatResourceStringIgnoreCodeAndKeyword("TaskHostAcquired_NullsFiltered", parameterName, nullCount),
+                helpKeyword: null,
+                senderName: taskName,
+                importance: MessageImportance.Normal));
```
[the remark this answers is not in the document]
Comment (JanProvaznik) [reply src/MSBuild/OutOfProcTaskAppDomainWrapperBase.cs:500]: the intent is for it to be visible at diagnostic verbostity not only in binlog

[the remark this answers is not in the document]
Comment (JanProvaznik) [reply src/MSBuild/OutOfProcTaskAppDomainWrapperBase.cs:500]: the point is that nulls are visible in binlog as task outputs

File: src/MSBuild/OutOfProcTaskAppDomainWrapperBase.cs [hunk 7]
```diff
@@ -451,5 +460,46 @@ private void LogErrorDelegate(string taskLocation, int taskLine, int taskColumn,
...
+        {
+            // Count nulls
+            int nullCount = 0;
+            foreach (string s in stringArray)
+            {
+                if (s == null)
+                {
+                    nullCount++;
+                }
+            }
```
[the remark this answers is not in the document]
Comment (JanProvaznik) [reply src/MSBuild/OutOfProcTaskAppDomainWrapperBase.cs:478]: the utility method call is guarded by `is string[] stringArray` which prevents that

File: src/MSBuild/OutOfProcTaskAppDomainWrapperBase.cs [hunk 8]
```diff
@@ -419,7 +419,16 @@ private OutOfProcTaskHostTaskResult InstantiateAndExecuteTask(
...
                     {
-                        finalParameterValues[value.Name] = value.GetValue(wrappedTask, null);
+                        object outputValue = value.GetValue(wrappedTask, null);
+
+                        // Filter null elements from string[] outputs to avoid crash.
+                        // See https://github.com/dotnet/msbuild/issues/13174
+                        if (outputValue is string[] stringArray)
+                        {
+                            outputValue = FilterNullsFromStringArray(stringArray, value.Name);
+                        }
```
[the remark this answers is not in the document]
Comment (JanProvaznik) [reply src/MSBuild/OutOfProcTaskAppDomainWrapperBase.cs:429]: updated description, integration test added
